### PR TITLE
Fixed Importing: non-default multiple value separator can't be used in custom multiple select attribute

### LINKED ISSUE #23156: Importing: non-default multiple value separator can't be used in custom multiple select attribute
### Preconditions (*)
Magento 2.4-develop

### Steps to reproduce (*)
1. Create a custom product attribute with type multi_select.
2. Create a catalog_product.csv that contains product data and at least 1 multiselect attribute with 2 options or set two option values to the product, save and make export
3. Separate options with custom separator, for example, **|||**
image
4. Go to the Magento admin panel -> `System` -> `Import`
5. Set `Multiple value separator` to **|||**
image
6. Choose your file and click the `Check Data` button

### Expected result (*)
1. Validation should pass without problems

### Actual result (*)
1. Validation fails and tells us we have incorrect values for our multiselect attribute
image

The code on https://github.com/magento/magento2/blob/2.3.1/app/code/Magento/CatalogImportExport/Model/Import/Product/Validator.php#L226

Fails to get the configured multiple value separator and pass it on in function $this->context->parseMultiselectValues.

NB I've actually imported the CSV programmaticaly and set parameter '_import_multiple_value_separator' as data in Magento\ImportExport\Model\ImportFactory->create().

Related tickets:

* CatalogImportExport multiselect values will fail validation with custom multiValueSeperator #6984
* Importing: non-default multiple value separator can't be used in additional_images column #11026

### PR DESCRIPTION
**Pull request by engcom-Charlie**
### Description (*)

Fixed issue, covered by the test, refactor test

### Fixed Issues (if relevant)

1. Fixes magento/magento2#23156

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by engcom-Charlie** [comment]
@magento run all tests

File: app/code/Magento/CatalogImportExport/Model/Import/Product/Validator.php [hunk 1]
```diff
@@ -227,7 +227,12 @@ public function isAttributeValid($attrCode, array $attrParams, array $rowData)
                 $valid = $this->validateOption($attrCode, $attrParams['options'], $rowData[$attrCode]);
                 break;
             case 'multiselect':
-                $values = $this->context->parseMultiselectValues($rowData[$attrCode]);
+                $separator = $this->context->getMultipleValueSeparator();
+                // added to prevent backward compatibility since before we didn't use separator param value
+                $values = $separator === ','
```
Comment (gabrieldagama) [inline app/code/Magento/CatalogImportExport/Model/Import/Product/Validator.php]: Let's replace this `','` with a private constant, probably call it `DEFAULT_GLOBAL_MULTI_VALUE_SEPARATOR`, to easier readability.

**Review by gabrieldagama — CHANGES_REQUESTED** [review]
Hi @engcom-Charlie, thanks for working on this. The changes look good, please follow a small suggestion.

Thanks!

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Review by gabrieldagama — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @gabrieldagama, thank you for the review.
[ENGCOM-8819](https://jira.corp.magento.com/browse/ENGCOM-8819) has been created to process this Pull Request
:eight_spoked_asterisk: @gabrieldagama, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Review by engcom-Bravo — CHANGES_REQUESTED** [review]
:x: QA Failed

After Checking Data, no errors appear. But after trying to **Import**, an _Undefined index_ error appears
undefined

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by therool** [comment]
Hello,
wouldn't it be just enough to change the method "parseMultipleValues" visibility to public in https://github.com/magento/magento2/blob/592c792621e6df4ad3d8de326cb5bd7a005583c7/app/code/Magento/CatalogImportExport/Model/Import/Product.php#L3132 so it can be called directly ("public function parseMultipleValues($labelRow)") and just alter the "multipleselect" attribute type validation on https://github.com/magento/magento2/blob/592c792621e6df4ad3d8de326cb5bd7a005583c7/app/code/Magento/CatalogImportExport/Model/Import/Product/Validator.php#L230 from "$values = $this->context->parseMultiselectValues($rowData[$attrCode]);" to "$values = $this->context->parseMultipleValues($rowData[$attrCode]);")

**Comment by ihor-sviziev** [comment]
@therool, unfortunately, that would be a backward-incompatible change.

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Please resolve conflicts

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8819](https://jira.corp.magento.com/browse/ENGCOM-8819) has been created to process this Pull Request
:eight_spoked_asterisk: @ihor-sviziev, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by ihor-sviziev** [comment]
Functional Tests B2B failure looks not related to changes in this PR.

**Comment by engcom-Foxtrot** [comment]
@magento run Functional Tests B2B

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Alfa** [comment]
**Before Test - Found no issue :** Tried reproducing the issue prior testing PR fixes in 2.4-develop; No issue noticed. Below are the steps executed!


1. Created the attribute with the name customMultiSelectProdAttri2 in the Admin : Stores > products > Create New Attribute. Below is the configured screenshot.
     image

2. Navigated to Store > Attribute Set > Add Attribute Set - Created the attribute set _custonAttributeSet1_ ; utilized above created attribute. Below is the screenshot used.
     image

3. Created a simple product using above created attribute type. Below is the product screenshot.

4. Navigated to Admin : System > Export > Select Entity Type - Products > Continue & Downloaded the CSV file.


**Note:** _Used the above steps to pre set the data and the same downloaded file is used to import again replacing with the new product data._
> **The downloaded file has additional_attributes by default:**
> additional_attributes
> customMultiSelectProdAttri2=option2|option3



5. Used the above downloaded csv file and replaced with new product data where set the additional info like:
additional_attributes
customMultiSelectProdAttri2=option2|option3


6. Navigated to the Admin: System > Import > Select Entity Type - Products; Import Behavior - Add/Update; Multiple Value Separator - ||| ; and then click check data button


**As per the reported issue, Should get the error while importing csv file; But no issue identified, check data succeeded as shown in the below screenshot.**

**Comment by engcom-Alfa** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Charlie** [comment]
Hi @engcom-Alfa,

I have tried to reproduce this issue on 2.4-develop branch and its still reproducible.

I can able to reproduce the issue by the steps mentioned in [your comments](https://github.com/magento/magento2/pull/31878#issuecomment-873899639). Below are the errors I am getting.

image

Please let me know if you need any other information.

Thank you!

**Comment by engcom-Alfa** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Alfa** [comment]
> Hi @engcom-Alfa,
> 
> I have tried to reproduce this issue on 2.4-develop branch and its still reproducible.
> 
> I can able to reproduce the issue by the steps mentioned in [your comments](https://github.com/magento/magento2/pull/31878#issuecomment-873899639). Below are the errors I am getting.
> 
> image
> 
> Please let me know if you need any other information.
> 
> Thank you!

Thanks @engcom-Charlie for re checking it. Yeah, I re tried today after your updates, able to reproduce using the same steps!

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Preconditions:**

1. Should have a Magento latest instance with sample data


**Manual testing scenario:**

Follow the same steps defined in [my last comment](https://github.com/magento/magento2/pull/31878#issuecomment-873899639).


**Before:** :heavy_multiplication_x: Data check to import the CSV file was getting failed
 image


**After:** :heavy_check_mark:  Data check to import the CSV file is getting successful.
 image

**Comment by engcom-Alfa** [comment]
Forwarding it for extended testing since a automation test job is getting failed continuously. Kind request to update the ticket with the fixes of failures @engcom-Charlie

**Comment by engcom-Charlie** [comment]
Hi @engcom-Alfa,

The failing test case is for Currency Converter API configuration. 

Its failing because its not able to retrieve the currency rate from other website api. None of the code changes made in this PR are responsible for this so moving this issue further.

Thank you!

**Pull request merged**